Improve code style in various examples

tshemsedinov opened the pull request:
#### Summary
- Remove async IIFE (now we have top-level await everywhere)
- Prefer `const` as for-await..of variable
- Optimize async iterator example
- Improve code style consistency

#### Motivation
Improve code examples to use MDN as a reference in my university lectures.

#### Metadata
- [ ] Adds a new document
- [ ] Rewrites (or significantly expands) a document
- [x] Fixes a typo, bug, or other error

tshemsedinov committed:
Improve code examples: for..of

- Remove async IIFE (now we have top-level await everywhere)
- Prefer `const` as for..of variable
- Improve code style consistency

hamishwillee commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
```diff
@@ -30,7 +30,7 @@ for await (const variable of iterable) {
   - : On each iteration a value of a different property is assigned to
     `variable`. `variable` may be declared
-    with `const`, `let`, or `var`.
+    with `const`, `let`, or `var` (prefer `const`).
```
I'm not sure about this addition. You prefer `const` if the variable is not to be modified in the loop, and `let` otherwise right?

tshemsedinov replied:
@hamishwillee what I want to say:
1. `var` is ok just for legacy code
2. `let` is allowed but not desirable because it allow mutations in loop body
3. `const` is preferable in almost all cases except some rare optimizations

wbamberg replied:
I don't think we should include "(prefer `const`)" here. We generally would recommend people to use const where they can, but I don't think there is sense in repeating this every time it comes up.

We should have code samples consistently using `const` though.

tshemsedinov replied:
Removed

tshemsedinov committed:
fixup! Improve code examples: for..of

tshemsedinov committed:
Improve code style consistency

- Improve variable naming
- Use constants instead of hardcoded magic numbers
- Prefer arrow functions
- Fix code alignment: add/remove spacing

tshemsedinov committed:
Prototype chain examples: improve code style

- Use `const` instead of `var`
- Fix code alignment: add/remove spacing

tshemsedinov committed:
Typed arrays: improve code style

Use const instead of let where applicable

tshemsedinov committed:
Merge branch 'main' into patch-1

wbamberg commented on files/en-us/learn/javascript/first_steps/arrays/index.md:
```diff
@@ -209,13 +209,9 @@ for (const bird of birds) {
-  return number * 2;
-}
-
+const double = (number) => number * 2;
```
It's intentional not to use arrow functions here, because we've not yet introduced them.

tshemsedinov replied:
```suggestion
function double(number) {
  return number * 2;
}
```

wbamberg commented on files/en-us/learn/javascript/first_steps/arrays/index.md:
```diff
@@ -224,13 +220,9 @@ We give a function to the `map()`, and `map()` calls the function once for each
-  return city.length > 8;
-}
-
+const isLong = (city) => city.length > 8;
```
It's intentional not to use arrow functions here, because we've not yet introduced them.

tshemsedinov replied:
```suggestion
function isLong(city) {
  return city.length > 8;
}
```

wbamberg commented on files/en-us/learn/javascript/first_steps/arrays/index.md:
```diff
@@ -122,76 +122,76 @@ console.log(birds.indexOf('Rabbit')) // -1
-const myArray = ['Manchester', 'Liverpool'];
-const removedItem = myArray.pop();
-console.log(removedItem);     // "Liverpool"
+const cities = ['Manchester', 'Liverpool'];
+const removedCities = cities.pop();
```
This would be `removedCity`, presumably, because it's a single item?

tshemsedinov replied:
```suggestion
const cities = ['Manchester', 'Liverpool'];
const removedCity = cities.pop();
```

wbamberg commented on files/en-us/learn/javascript/first_steps/arrays/index.md:
```diff
@@ -355,16 +347,16 @@ function updateCode() {
 }
 
-reset.addEventListener('click', function() {
+reset.addEventListener('click', () => {
```
This is internal code (not shown to users) so it's OK to use arrow functions.

wbamberg commented on files/en-us/web/javascript/inheritance_and_the_prototype_chain/index.md:
````diff
@@ -82,7 +84,7 @@ JavaScript does not have "methods" in the form that class-based languages define
 
 ```js
-var o = {
+const o = {
````
+1 to var->const changes!

wbamberg commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
````diff
@@ -41,6 +41,8 @@ for await (const variable of iterable) {
 You can also iterate over an object that explicitly implements async iterable protocol:
 
 ```js
+const LIMIT = 3;
````
this doesn't seem to be used?

tshemsedinov replied:
Oh, It looks like I dropped part of this example rebasing... I'll add it back in following commit

wbamberg commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
```diff
@@ -60,12 +62,9 @@ const asyncIterable = {
-   }
-})();
-
+for await (const num of asyncIterable) {
```
Putting this in the console (Firefox 98) it doesn't work:

> Uncaught SyntaxError: for await (... of ...) is only valid in async functions and async generators

...while the original IIFE version does.

wbamberg reviewed (changes requested):
Thanks for this PR, @tshemsedinov ! Some of these style changes are important IMO, especially `var`->`const`. Some of them I think are a bit more subjective, but on the whole I agree with pretty much all of them. I had a few complaints, but if we can address them I'm happy to merge this.

I do think, though, that for MDN we really need to be addressing code style issues with tooling across the whole site, rather than a bit at a time through manual PRs. The site is just too big, and our volume of PRs is just too high, for manual fixes to be sustainable.

We have [an open issue about enforcing Prettier (and hopefully ESLint) for our code samples](https://github.com/mdn/content/issues/10402), and I hope we will get to that soonish.

If there are particular areas of MDN for which you want to update the code style, it might be worth talking to us about the scope of that (maybe in a [discussion](https://github.com/mdn/content/discussions)?) so we can figure out how to resource reviews for it. But I don't think we can update the entire site in this way.

I don't want to discourage you because I think these are good improvements, we just need to figure out an efficient way to make them.

tshemsedinov commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
```diff
@@ -30,7 +30,7 @@ for await (const variable of iterable) {
   - : On each iteration a value of a different property is assigned to
     `variable`. `variable` may be declared
-    with `const`, `let`, or `var`.
+    with `const`, `let`, or `var` (prefer `const`).
```
```suggestion
    with `const`, `let`, or `var`.
```

tshemsedinov commented on files/en-us/learn/javascript/first_steps/arrays/index.md:
```diff
@@ -122,76 +122,76 @@ console.log(birds.indexOf('Rabbit')) // -1
-console.log(removedItem);     // "Liverpool"
+const cities = ['Manchester', 'Liverpool'];
+const removedCities = cities.pop();
+console.log(removedCities);   // "Liverpool"
```
```suggestion
console.log(removedCity);   // "Liverpool"
```

tshemsedinov commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
```diff
@@ -84,11 +83,9 @@ async function* asyncGenerator() {
-    console.log(num);
-  }
-})();
+for await (const num of asyncGenerator()) {
+  console.log({ num });
+}
```
```suggestion
(async () => {
  for await (const num of asyncGenerator()) {
    console.log({ num });
  }
})();
```

tshemsedinov commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
```diff
@@ -148,11 +144,9 @@ function* generator() {
-    console.log(num);
-  }
-})();
+for await (const num of generator()) {
+  console.log(num);
+}
```
```suggestion
(async () => {
  for await (const num of generator()) {
    console.log(num);
  }
})();
```

tshemsedinov committed:
Apply suggestions from code review

tshemsedinov commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
```diff
@@ -187,19 +185,17 @@ function* generatorWithRejectedPromises() {
-    }
-  } catch (e) {
-    console.log('caught', e)
+try {
+  for await (const num of generatorWithRejectedPromises()) {
+    console.log(num);
```
```suggestion
(async () => {
  try {
    for await (const num of generatorWithRejectedPromises()) {
      console.log(num);
```

tshemsedinov commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
```diff
@@ -187,19 +185,17 @@ function* generatorWithRejectedPromises() {
+    console.log(num);
   }
-})();
+} catch (e) {
+  console.log('caught', e);
+}
```
```suggestion
  } catch (e) {
    console.log('caught', e);
  }
})();
```

tshemsedinov commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
```diff
@@ -229,15 +225,13 @@ appropriate form of the loop, `for await...of` for the async generator and
+    console.log(await numOrPromise);
   }
-})()
+} catch (e) {
+  console.log('caught', e);
+}
```
```suggestion
  } catch (e) {
    console.log('caught', e);
  }
})();
```

tshemsedinov committed:
Apply suggestions from code review

tshemsedinov committed:
Fixup

tshemsedinov commented on files/en-us/web/javascript/reference/statements/for-await...of/index.md:
```diff
@@ -41,16 +41,16 @@ for await (const variable of iterable) {
-        }
-
-        return Promise.resolve({ done: true });
+        const done = this.i === LIMIT;
+        const value = done ? undefined : this.i++;
+        return Promise.resolve({ value, done });
```
```suggestion
        const done = i === LIMIT;
        const value = done ? undefined : i++;
        return Promise.resolve({ value, done });
```

tshemsedinov committed:
Apply suggestions from code review

tshemsedinov committed:
Fixup

tshemsedinov committed:
Fixup

tshemsedinov commented:
@wbamberg Thanks! Everything resolved, lost part of code with `asyncIterable` example returned. In future it will be good to have automated tools and concentrate on more complex code improvements, like we have in mentioned `asyncIterable` example: https://github.com/tshemsedinov/content/blob/patch-1/files/en-us/web/javascript/reference/statements/for-await...of/index.md#iterating-over-async-iterables

wbamberg reviewed (approved):
Looks great! Thanks for the updates.
